### ISSUE TITLE
[CSS REWRITE] [BIG UPDATE] Change how to process CSS 

### ISSUE THREAD

**Issue by nvh95** [nvh95/jest-preview#252]
### Current approach to process CSS
- Currently, we use Jest Code Transformer to transform CSS. Except Pure CSS, we transform different CSS languages such as Sass, CSS Modules, TailwindCSS, PostCSS... then embed them directly the **output CSS** to the `document.head` of `jsdom`. "Jest Preview server" will serve `jsdom` as real DOM, so we can see the CSS on the "Jest Preview Dashboard". However, this approach encounters a number of problems
  - (The most annoying one) Jest Code Transformer does not support async actions by default. That means the code must be in "SYNC" style. It's very hard to do some asynchronous actions in a syncchronous manner. Sometimes, I could find the sync API to use. Otherwise, I ended up with spawning a new node process, `console.log` the output the stdout, then reading/ parsing to get the returned value. This is a tricky one and does not scale, and I considered it as a unscale code.
  - After #241, the time processing CSS file takes a long time. For example, processing a CSS file in a project with PostCSS configuration would take 0.3 seconds on Macbook Air M1. If the user test case contains 10 CSS files, the total time would be 3 seconds. I don't like the idea of waiting 3 seconds before a user's test can run. To be fair, this is not a critical problem, since Jest will cache the computation result, so it's likely you only need to wait once.

### Is there any better solution

Yes. But first, let me describe how I come up with that idea.

A few days ago, I was working on #230. The issue is that when a CSS file that contains an `@import` statement, `jest-preview` could not load the content of the imported file:

```css
// index.css. This file is imported to the setupTest.js
@import "base.css";
@import "components.css";
@import "utils.css";
```

If you are curious, I handled this in #248, by inline the imported CSS, using (`postcss-import`). But the thing is, when opening Jest Preview Dashboard my `index.css` will automatically load CSS files that are imported. (It's a [CSS standard](https://www.w3schools.com/cssref/pr_import_rule.asp))

![](https://user-images.githubusercontent.com/8603085/186130322-7a6512ae-e843-4162-8b7c-825925cc5b23.png)

This opens a new horizon for me to resolve all those issues above. What if instead of processing CSS at Jest Code Transformation, the "Jest Preview Dashboard" will process the CSS? I am in the middle of dinner with my family and I have to stop to think and draft the solution. This is briefly how it would works:
- Jest Preview does not process CSS anymore. It just inject the URL to the `jsdom`'s `document.head`. Like this

```js
```

**Comment by nvh95** [comment]
Hey @ntt261298, please see my note above on a new approach to handle CSS. It would be great if you can give some of your thoughts. Thanks.

**Comment by ntt261298** [comment]
I think this approach is really promising. (Reduce the test running time, Be able to handle async tasks so we can scale easier in the future,…)
I still have some thoughts about this approach:
  - The time for the browser to fully load the CSS of the preview might be longer than current (But I think reducing the test running time is more important).
  - We might still need to transform CSS module in the transformer. As I understand, we still need to do the expensive task `postcssrc()` for every test. What if we process the CSS module in the Jest Preview Dashboard? Then it might take a much longer time for the browser to load CSS fully. (Not sure if the user has enough patience to wait for that). I think this is a trade-off.
  - Reducing the time of running `postcssrc()` might be the problem we need to solve for both approaches (current and new), but it’s probably the really hard one.
  
To summarize, I think it’s feasible to implement the new approach and will bring more pros than cons.

**Comment by nvh95** [comment]
[Just a note] This can fix #260

**Comment by szamanr** [comment]
hi, is this on the roadmap? i would like to reduce the runtime in CI or make it only parse css based on some env variable. looks like this rewrite would solve that.
